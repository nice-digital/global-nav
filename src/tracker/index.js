export const eventTimeout = 2000;

export const eventName = "GlobalNav";
export const defaultEventCategory = "TopHat and footer";
export const headerClickEventAction = "Tophat click";
export const footerClickEventAction = "Footer click";

export const ensureDataLayer = function () {
	window.dataLayer = window.dataLayer || [];
};

/**
 * Pushes an event object to the GTM dataLayer
 *
 * @param {string} category The event category
 * @param {string} action The event action
 * @param {string} label The event label
 * @param {Number} value The optional event value
 * @param {string} destinationUrl The optional link destination
 * @param {Function} callback The callback function to be called when the GTM event has fired
 */
export const trackEvent = function (
	category,
	action,
	label,
	value = null,
	destinationUrl = null,
	callback = null
) {
	ensureDataLayer();

	const dataLayerEvent = {
		event: eventName,
		eventCategory: category,
		eventAction: action,
		eventLabel: label,
	};

	if (value) {
		dataLayerEvent.eventValue = value;
	}

	if (destinationUrl) {
		dataLayerEvent.destinationUrl = destinationUrl;
	}

	if (callback) {
		let callbackTimeout;
		const wrappedCallback = function () {
			clearTimeout(callbackTimeout);
			callback();
		};
		callbackTimeout = setTimeout(wrappedCallback, eventTimeout);
		dataLayerEvent.eventCallback = wrappedCallback;
		dataLayerEvent.eventTimeout = eventTimeout;
	}

	window.dataLayer.push(dataLayerEvent);
};
