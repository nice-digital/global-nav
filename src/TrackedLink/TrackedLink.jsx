import React, { Component } from "react";
import PropTypes from "prop-types";

import { defaultEventCategory, trackEvent } from "./../tracker";

export default class TrackedLink extends Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
	}

	handleClick(e) {
		e.preventDefault();

		let { eventCategory, eventAction, eventLabel } = this.props;

		const href = e.currentTarget.getAttribute("href");

		trackEvent(
			eventCategory,
			eventAction,
			eventLabel || e.currentTarget.textContent || e.currentTarget.innerText,
			null,
			href,
			(function () {
				window.location.assign(href);
			})()
		);
	}

	render() {
		/* eslint-disable no-unused-vars */
		const {
			children,
			href,
			eventCategory,
			eventAction,
			eventLabel,
			eventValue,
			destinationUrl,
			...props
		} = this.props;
		/* eslint-enable no-unused-vars */

		return (
			<a href={href} onClick={this.handleClick} {...props}>
				{children}
			</a>
		);
	}
}

TrackedLink.propTypes = {
	href: PropTypes.string.isRequired,
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node,
	]).isRequired,
	eventCategory: PropTypes.string,
	eventAction: PropTypes.string.isRequired,
	eventLabel: PropTypes.string,
	eventValue: PropTypes.number,
	destinationUrl: PropTypes.string,
};

TrackedLink.defaultProps = {
	eventCategory: defaultEventCategory,
};
