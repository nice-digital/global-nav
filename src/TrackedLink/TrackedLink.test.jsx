import React from "react";
import TrackedLink from "./TrackedLink";
import { render, createEvent, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { eventName, defaultEventCategory } from "../tracker";

describe("TrackedLink", () => {
	beforeEach(() => {
		window.dataLayer = [];

		window.location = { ...window.location, assign: jest.fn() };
	});

	afterAll(() => {
		// Cleanup
		delete window.dataLayer;

		window.location.assign.mockClear();
	});

	it("Matches snapshot", () => {
		const { container } = render(
			<TrackedLink href="/a-url" eventAction="Action" className="className">
				Some text
			</TrackedLink>
		);
		expect(container).toMatchSnapshot();
	});

	it("should send event to the dataLayer with navigating event callback on link click", async () => {
		const href = "https://www.nice.org.uk/a-url",
			eventName = "GlobalNav",
			eventAction = "Test action",
			eventLabel = "Test label";

		const { getByRole } = render(
				<TrackedLink
					eventAction={eventAction}
					eventLabel={eventLabel}
					href={href}
				>
					Some text
				</TrackedLink>
			),
			link = getByRole("link"),
			clickEvent = createEvent.click(link);

		fireEvent(link, clickEvent);

		expect(window.dataLayer).toStrictEqual([
			{
				event: eventName,
				eventCategory: defaultEventCategory,
				eventAction: eventAction,
				eventLabel: eventLabel,
				destinationUrl: href,
			},
		]);

		expect(clickEvent.defaultPrevented).toBe(true);
		expect(window.location.assign).toHaveBeenCalledWith(href);
	});

	it("should use default event category when none provided", async () => {
		const href = "https://www.nice.org.uk/a-url",
			eventAction = "Test action",
			eventLabel = "Some text";

		const { getByRole } = render(
				<TrackedLink
					href={href}
					eventAction={eventAction}
					eventLabel={eventLabel}
				>
					Some text
				</TrackedLink>
			),
			link = getByRole("link"),
			user = userEvent.setup();

		await user.click(link);

		expect(window.dataLayer).toStrictEqual([
			{
				event: eventName,
				eventCategory: defaultEventCategory,
				eventAction: eventAction,
				eventLabel: eventLabel,
				destinationUrl: href,
			},
		]);
	});

	it("should use link textContent property for event label when none provided", async () => {
		const eventAction = "Test action",
			href = "https://www.nice.org.uk/a-url",
			textContent = "Some text content";

		const { getByRole } = render(
				<TrackedLink href={href} eventAction={eventAction}>
					{textContent}
				</TrackedLink>
			),
			link = getByRole("link"),
			user = userEvent.setup();

		await user.click(link);

		expect(window.dataLayer).toStrictEqual([
			{
				event: eventName,
				eventCategory: defaultEventCategory,
				eventAction: eventAction,
				eventLabel: textContent,
				destinationUrl: href,
			},
		]);
	});
});
