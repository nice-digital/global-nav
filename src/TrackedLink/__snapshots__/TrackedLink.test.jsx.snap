// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackedLink Matches snapshot 1`] = `
<div>
  <a
    class="className"
    href="/a-url"
  >
    Some text
  </a>
</div>
`;
